### PR TITLE
Problem: Missing signers in transaction API

### LINKED ISSUE #438: Problem: Missing signers in transaction API
Transaction events already have the signers data, we just need to store it in the projection

Events:
https://github.com/crypto-com/chain-indexing/blob/master/usecase/event/transaction_created.go#L24
https://github.com/crypto-com/chain-indexing/blob/master/usecase/event/transaction_failed.go#L24

Projection
https://github.com/crypto-com/chain-indexing/blob/master/projection/transaction/transaction.go#L74

### PR DESCRIPTION
**Pull request by davcrypto**
Solution: fixed #438 by adding signers field into transaction struct

Modified response would be like:
```
        {
            "blockHeight": 13618,
            "blockHash": "625B78B0C643021B9C917BFCC2E8CDE8F9A39341DFCB538EFB5A86ED26B5B87F",
            ...,
            "signers": [
                {
                    "type": "/cosmos.crypto.secp256k1.PubKey",
                    "isMultiSig": false,
                    "pubkeys": [
                        "Ah5t0VGq6AIcsvPYT1lYjKq6thZE3kpaPtbolj61vgcU"
                    ],
                    "accountSequence": 3,
                    "address": "tcro1f5rds2xlqswhpxynp6v62k4kz328fxd00059w4"
                }
            ]
        }
```

### CODE REVIEW

File: projection/transaction/transaction.go [hunk 1]
```diff
@@ -72,25 +72,41 @@ func (projection *Transaction) HandleEvents(height int64, events []event_entity.
...
-				Fee:           transactionCreatedEvent.Fee,
-				FeePayer:      transactionCreatedEvent.FeePayer,
-				FeeGranter:    transactionCreatedEvent.FeeGranter,
-				GasWanted:     transactionCreatedEvent.GasWanted,
-				GasUsed:       transactionCreatedEvent.GasUsed,
-				Memo:          transactionCreatedEvent.Memo,
-				TimeoutHeight: transactionCreatedEvent.TimeoutHeight,
-				Messages:      make([]transaction_view.TransactionRowMessage, 0),
-			})
+			if transactionCreatedEvent != nil {
```
Comment (calvinaco) [inline projection/transaction/transaction.go:75]: Why do we need a nil check here?

File: migrations/20210805154736_alter_view_transactions_add_signer.up.sql [hunk 2]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE view_transactions
+    ADD signers JSONB NOT NULL;
```
Comment (calvinaco) [inline migrations/20210805154736_alter_view_transactions_add_signer.up.sql:2]: For existing transactions in the table, will `NOT NULL` fails them? Perhaps put an empty array as default value ?

File: projection/transaction/view/transactions.go [hunk 3]
```diff
@@ -72,6 +73,13 @@ func (transactionsView *BlockTransactions) InsertAll(transactions []TransactionR
 			)
 		}
 
+		var signersJSON string
+		if signersJSON, marshalErr = jsoniter.MarshalToString(transaction.Signers); marshalErr != nil {
+			return fmt.Errorf(
+				"error JSON marshalling block transation fee for insertion: %v: %w", marshalErr, rdb.ErrBuildSQLStmt,
```
Comment (calvinaco) [inline projection/transaction/view/transactions.go:79]:
```suggestion
				"error JSON marshalling block transaction signer for insertion: %v: %w", marshalErr, rdb.ErrBuildSQLStmt,
```

File: projection/transaction/view/transactions.go [hunk 4]
```diff
@@ -151,6 +161,11 @@ func (transactionsView *BlockTransactions) Insert(transaction *TransactionRow) e
 		return fmt.Errorf("error JSON marshalling block transation fee for insertion: %v: %w", err, rdb.ErrBuildSQLStmt)
 	}
 
+	var signersJSON string
+	if signersJSON, err = jsoniter.MarshalToString(transaction.Signers); err != nil {
+		return fmt.Errorf("error JSON marshalling block transation fee for insertion: %v: %w", err, rdb.ErrBuildSQLStmt)
```
Comment (calvinaco) [inline projection/transaction/view/transactions.go:166]:
```suggestion
		return fmt.Errorf("error JSON marshalling block transaction signers for insertion: %v: %w", err, rdb.ErrBuildSQLStmt)
```

File: projection/transaction/view/transactions.go [hunk 5]
```diff
@@ -254,6 +273,12 @@ func (transactionsView *BlockTransactions) FindByHash(txHash string) (*Transacti
 	}
 	transaction.Messages = messages
 
+	var signers []TransactionRowSigner
+	if unmarshalErr := jsoniter.UnmarshalFromString(*signersJSON, &signers); unmarshalErr != nil {
+		return nil, fmt.Errorf("error unmarshalling transaction messages JSON: %v: %w", unmarshalErr, rdb.ErrQuery)
```
Comment (calvinaco) [inline projection/transaction/view/transactions.go:278]:
```suggestion
		return nil, fmt.Errorf("error unmarshalling transaction signers JSON: %v: %w", unmarshalErr, rdb.ErrQuery)
```

File: projection/transaction/view/transactions.go [hunk 6]
```diff
@@ -367,6 +395,12 @@ func (transactionsView *BlockTransactions) List(
 		}
 		transaction.Messages = messages
 
+		var signers []TransactionRowSigner
+		if unmarshalErr := jsoniter.UnmarshalFromString(*signersJSON, &signers); unmarshalErr != nil {
+			return nil, nil, fmt.Errorf("error unmarshalling transaction messages JSON: %v: %w", unmarshalErr, rdb.ErrQuery)
```
Comment (calvinaco) [inline projection/transaction/view/transactions.go:400]:
```suggestion
			return nil, nil, fmt.Errorf("error unmarshalling transaction signers JSON: %v: %w", unmarshalErr, rdb.ErrQuery)
```

**Review by calvinaco — COMMENTED** [review]
Thanks for your first PR contribution @davcrypto . Overall it's looking good. I left a few comments below.

File: projection/transaction/transaction.go (hunk 1 above)
Comment (davcrypto) [reply projection/transaction/transaction.go]: Should be false positive on lint, already removed

File: migrations/20210805154736_alter_view_transactions_add_signer.up.sql (hunk 2 above)
Comment (davcrypto) [reply migrations/20210805154736_alter_view_transactions_add_signer.up.sql:2]: Running in testnet env has no issue on this. But true, more safety to put a default value here. Will do it.

**Review by calvinaco — APPROVED** [review]
LGTM

**Pull request merged**